### PR TITLE
Add an example for AMP cache. Fixes #87

### LINKED ISSUE #87: AMP Cache example
Create an example with AMP cache

### PR DESCRIPTION
**Pull request by kul3r4**

### CODE REVIEW

**Comment by kul3r4** [comment]
I signed it!

2016-03-18 12:21 GMT+00:00 googlebot <notifications@github.com>:

> Thanks for your pull request. It looks like this may be your first
> contribution to a Google open source project. Before we can look at your
> pull request, you'll need to sign a Contributor License Agreement (CLA).
>
> [image: :memo:] *Please visit https://cla.developers.google.com/
> <https://cla.developers.google.com/> to sign.*
>
> Once you've signed, please reply here (e.g. I signed it!) and we'll
> verify. Thanks.
> ------------------------------
>
>    - If you've already signed a CLA, it's possible we don't have your
>    GitHub username or you're using a different email address. Check your
>    existing CLA data <https://cla.developers.google.com/clas> and verify
>    that your email is set on your git commits
>    <https://help.github.com/articles/setting-your-email-in-git/>.
>    - If you signed the CLA as a corporation, please let us know the
>    company's name.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/ampproject/amp-by-example/pull/104#issuecomment-198328121>
>



-- 
Chiara Chiappini

File: src/30_Advanced/AMP_cache.html [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    <!-- #### GET Arguments -->
+    <!--
+      GET arguments are supported when using AMP cache.
+    -->
+
+    <a href="https://cdn.ampproject.org/c/www.standard.co.uk/showbiz/celebrity-news/harry-styles-officially-added-to-cast-list-for-christopher-nolan-s-dunkirk-a3204886.html?amp">Link containing a GET argument.</a>
+    
+    <!-- #### Redirect -->
+    <!--
+      Redirect is supported when using AMP cache.
```
Comment (sebastianbenz) [inline src/30_Advanced/AMP_cache.html]: Can you explain this a bit more.

File: src/30_Advanced/AMP_cache.html [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+  <!--
+    Prepend `cdn.ampproject.org/c/s` to AMP Urls. 
+
+  -->
+    <a href="https://ampbyexample.com/components/amp-video/">Link with AMP HTTPS URL</a>
+    <a href="https://cdn.ampproject.org/c/s/ampbyexample.com/components/amp-video/">Link with AMP HTTPS URL accessing the cache</a>
+    
+    <!-- #### GET Arguments -->
+    <!--
+      GET arguments are supported when using AMP cache.
```
Comment (sebastianbenz) [inline src/30_Advanced/AMP_cache.html]: Can you explain this a bit more - what does support mean?

File: src/30_Advanced/AMP_cache.html [hunk 3]
```diff
@@ -0,0 +1,56 @@
+<!--
+  #### Introduction
+
+  Google provides a cache that clients can use for free to display AMP files. This gives you two choices to display an AMP file in your app using either:
```
Comment (sebastianbenz) [inline src/30_Advanced/AMP_cache.html]: We have to make the intro more generic - the bit you've used is specific for apps.

File: src/30_Advanced/AMP_cache.html [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    
+    <!-- #### Redirect -->
+    <!--
+      Redirect is supported when using AMP cache.
+    -->
+    <a href="https://cdn.ampproject.org/c/s/ampbyexample.com/amp-img.html">Link which will result in a redirect to https://ampbyexample.com/components/amp-img/.</a>
+    
+    <!-- #### Not Found Page -->
+    <!--
+      When a page is not found in the amp cache, 404 will be returned.
```
Comment (sebastianbenz) [inline src/30_Advanced/AMP_cache.html]: When a page is not found in the amp cache, the AMP Cache will show an error page and return a 404 error.

**Comment by kul3r4** [comment]
I added some additional explanations, PTAL

**Comment by kul3r4** [comment]
@sebastianbenz  PTAL

File: server.go [hunk 5]
```diff
@@ -49,6 +49,7 @@ var REDIRECTS [18][2]string = [18][2]string{
 func init() {
 	RedirectLegacyExamples()
 	http.HandleFunc("/g", getParameterDemoHandler)
+	http.HandleFunc("/returnInternalServerError", returnCode500)
```
Comment (sebastianbenz) [inline server.go]: s/returnInternalServerError/InternalServerErrorHandler/g

File: server.go [hunk 6]
```diff
@@ -83,3 +84,8 @@ func RedirectDomain(h http.Handler) http.Handler {
 		h.ServeHTTP(w, r)
 	})
 }
+
+ func returnCode500(w http.ResponseWriter, r *http.Request) {
+ 	w.WriteHeader(http.StatusInternalServerError)
+ 	w.Write([]byte("Internal Server Error Status retuned"))
```
Comment (sebastianbenz) [inline server.go]: s/"Internal Server Error Status retuned"/"Internal Server Error"/g

File: static/invalid_amp_page.html [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+      <meta charset="utf-8">
+
+  <meta name="viewport" content="width=device-width,minimum-scale=1,initial-scale=1">
+  <style amp-boilerplate>body{-webkit-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-moz-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-ms-animation:-amp-start 8s steps(1,end) 0s 1 normal both;animation:-amp-start 8s steps(1,end) 0s 1 normal both}@-webkit-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-moz-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-ms-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-o-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}</style><noscript><style amp-boilerplate>body{-webkit-animation:none;-moz-animation:none;-ms-animation:none;animation:none}</style></noscript>
+  <script async src="https://cdn.ampproject.org/v0.js"></script>
+
+    
+  </head>
+  <body>
+    <p>Invalid AMP page</p>
```
Comment (sebastianbenz) [inline static/invalid_amp_page.html]: s/Invalid AMP page/This page is intentionally invalid AMP/g

**Comment by sebastianbenz** [comment]
* Can you include our test page for the get parameter sample: https://ampbyexample.com/g?value=Hello%20World The link will work once you remove the line including "template" in app.yaml. 
* Can you add a warning box to the redirect sample? You need to define the class `warning` which adds  a background color to the card, e.g. #F8BBD0.
```
<p class="card warning">If you move the location of the AMP Files on your server, make sure to set up a redirect from the old location to the new one. </p>
```

**Comment by kul3r4** [comment]
PTAL

File: src/30_Advanced/Using_the_Google_AMP_Cache.html [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+  }
+    </style>
+</head>
+<body>
+  <!-- #### AMP Cache URL Format -->
+  <!--
+    Prepend `cdn.ampproject.org/c/` to AMP HTTP Urls. 
+  -->
+    <a href="http://www.bbc.co.uk/news/amp/35838735">Link with AMP HTTP URL</a>
+    <a href="https://cdn.ampproject.org/c/www.bbc.co.uk/news/amp/35838735">Link with AMP HTTP URL accessing the cache</a>
```
Comment (sebastianbenz) [inline src/30_Advanced/Using_the_Google_AMP_Cache.html]: s/Link with AMP HTTP URL/Link to AMP File/
s/Link with AMP HTTP URL accessing the cache/Link to AMP File via AMP Cache/

File: src/30_Advanced/Using_the_Google_AMP_Cache.html [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+  <!--
+    Prepend `cdn.ampproject.org/c/s` to AMP HTTPS Urls. 
+
+  -->
+    <a href="https://ampbyexample.com/components/amp-video/">Link with AMP HTTPS URL</a>
+    <a href="https://cdn.ampproject.org/c/s/ampbyexample.com/components/amp-video/">Link with AMP HTTPS URL accessing the cache</a>
```
Comment (sebastianbenz) [inline src/30_Advanced/Using_the_Google_AMP_Cache.html]: s/Link with AMP HTTP URL/Link to AMP File (HTTPS)/
s/Link with AMP HTTP URL accessing the cache/Link to AMP File via AMP Cache (HTTPS)/

File: src/30_Advanced/Using_the_Google_AMP_Cache.html [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+  <script async src="https://cdn.ampproject.org/v0.js"></script>
+  <style amp-custom>
+  .warning {
+    background-color: #F8BBD0;
+  }
+    </style>
+</head>
+<body>
+  <!-- #### AMP Cache URL Format -->
+  <!--
```
Comment (sebastianbenz) [inline src/30_Advanced/Using_the_Google_AMP_Cache.html]: Add "Loading your AMP pages via the Google AMP Cache is easy. The AMP Cache URL is composed based on the source URL’s scheme (HTTP or HTTPS):"

File: src/30_Advanced/Using_the_Google_AMP_Cache.html [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+
+  <!--
+    Prepend `cdn.ampproject.org/c/s` to AMP HTTPS Urls. 
+
+  -->
+    <a href="https://ampbyexample.com/components/amp-video/">Link with AMP HTTPS URL</a>
+    <a href="https://cdn.ampproject.org/c/s/ampbyexample.com/components/amp-video/">Link with AMP HTTPS URL accessing the cache</a>
+    
+    <!--
+      If URLs you are using are using GET parameters for implementing specific logics, you can still specify them in the new URL created for using the AMP cache.
```
Comment (sebastianbenz) [inline src/30_Advanced/Using_the_Google_AMP_Cache.html]: It is OK for URLs to include parameters in the query string, simply include these in the AMP Cache URL as well.

File: src/30_Advanced/Using_the_Google_AMP_Cache.html [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+    -->
+
+    <a href="https://cdn.ampproject.org/c/s/ampbyexample.com/g?value=Hello%20World">Link containing a GET argument.</a>
+    
+    <!-- #### Redirect & Error Handling -->
+    <!--
+      If your server is redirecting a particular URL to another URL, this will still work when using AMP cache.
```
Comment (sebastianbenz) [inline src/30_Advanced/Using_the_Google_AMP_Cache.html]: sug: The AMP Cache follows redirects when resolving AMP URLs.

File: src/30_Advanced/Using_the_Google_AMP_Cache.html [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+    -->
+    <a href="https://cdn.ampproject.org/c/s/ampbyexample.com/amp-img.html">Link which will result in a redirect to https://ampbyexample.com/components/amp-img/.</a>
```
Comment (sebastianbenz) [inline src/30_Advanced/Using_the_Google_AMP_Cache.html]: s/Link which will result in a redirect to/Link redirecting to/

File: src/30_Advanced/Using_the_Google_AMP_Cache.html [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+    <p class="card warning">If you move the location of the AMP Files on your server, make sure to set up a redirect from the old location to the new one. </p>
```
Comment (sebastianbenz) [inline src/30_Advanced/Using_the_Google_AMP_Cache.html]: can you add a css class for warning. So my comment for the previous commit for a suggested color.

File: src/30_Advanced/Using_the_Google_AMP_Cache.html [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+
+    
+    <!--
+      When a page is not found in the amp cache, the AMP Cache will show an error page and return a 404 error.
+    -->
+    <a href="https://cdn.ampproject.org/c/s/ampbyexample.com/not-found">Link to a non existing page</a>
+
+    <!--
+      When a page is invalid, the AMP Cache will return a 404 error.
```
Comment (sebastianbenz) [inline src/30_Advanced/Using_the_Google_AMP_Cache.html]: s/invalid/invalid AMP/

File: src/30_Advanced/Using_the_Google_AMP_Cache.html [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+
+  -->
+    <a href="https://ampbyexample.com/components/amp-video/">Link with AMP HTTPS URL</a>
+    <a href="https://cdn.ampproject.org/c/s/ampbyexample.com/components/amp-video/">Link with AMP HTTPS URL accessing the cache</a>
+    
+    <!--
+      If URLs you are using are using GET parameters for implementing specific logics, you can still specify them in the new URL created for using the AMP cache.
+    -->
+
+    <a href="https://cdn.ampproject.org/c/s/ampbyexample.com/g?value=Hello%20World">Link containing a GET argument.</a>
```
Comment (sebastianbenz) [inline src/30_Advanced/Using_the_Google_AMP_Cache.html]: s/Link containing a GET argument/Link with GET Parameter/

File: src/30_Advanced/Using_the_Google_AMP_Cache.html [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+    <!-- #### Redirect & Error Handling -->
+    <!--
+      If your server is redirecting a particular URL to another URL, this will still work when using AMP cache.
+    -->
+    <a href="https://cdn.ampproject.org/c/s/ampbyexample.com/amp-img.html">Link which will result in a redirect to https://ampbyexample.com/components/amp-img/.</a>
+    <p class="card warning">If you move the location of the AMP Files on your server, make sure to set up a redirect from the old location to the new one. </p>
+
+    
+    <!--
+      When a page is not found in the amp cache, the AMP Cache will show an error page and return a 404 error.
```
Comment (sebastianbenz) [inline src/30_Advanced/Using_the_Google_AMP_Cache.html]: s/amp cache/AMP Cache/
s/the AMP Cache/it/

**Comment by sebastianbenz** [comment]
I've added a few more comments.

**Comment by sebastianbenz** [comment]
Can you also please s/Url/URL/g

File: src/30_Advanced/Using_the_Google_AMP_Cache.html (hunk 14 above)
Comment (kul3r4) [inline src/30_Advanced/Using_the_Google_AMP_Cache.html]: isn't the class I already any good?
  <style amp-custom>
  .warning {
    background-color: #F8BBD0;
  }
    </style>

**Comment by kul3r4** [comment]
Changed background color for warning and added some margin.
PTAL.

**Comment by sebastianbenz** [comment]
@kul3r4 I've updated the PR with our changes. PTAL

**Pull request merged**